node/node.go

```diff
@@ -0,0 +1,452 @@
+	select {
+	case err := <-errCh:
+		if err != nil {
+			return nil, nil, err
+		}
+	default:
+		go func() {
+			if err := <-errCh; err != nil {
+				log.Println("WARNING: consensus initialization returned an error:", err)
+			}
+		}()
+	}
+	tpoolDir := filepath.Join(dir, "transactionpool")
+	if err := os.MkdirAll(tpoolDir, 0700); err != nil {
+		return nil, nil, err
+	}
+	tp, err := transactionpool.New(cm, g, tpoolDir)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	walletDir := filepath.Join(dir, "wallet")
+	if err := os.MkdirAll(walletDir, 0700); err != nil {
+		return nil, nil, err
+	}
+	walletAddr := wallet.StandardAddress(walletKey.PublicKey())
+	ws, ccid, err := stores.NewJSONWalletStore(walletDir, walletAddr)
+	if err != nil {
+		return nil, nil, err
+	} else if err := cm.ConsensusSetSubscribe(ws, ccid, nil); err != nil {
+		return nil, nil, err
+	}
+	w := wallet.NewSingleAddressWallet(walletKey, ws)
+
+	dbDir := filepath.Join(dir, "db")
+	if err := os.MkdirAll(dbDir, 0700); err != nil {
+		return nil, nil, err
+	}
+	dbConn := stores.NewSQLiteConnection(filepath.Join(dbDir, "db.sqlite"))
+
+	hostdbDir := filepath.Join(dir, "hostdb")
+	if err := os.MkdirAll(hostdbDir, 0700); err != nil {
+		return nil, nil, err
+	}
+	sqlStore, ccid, err := stores.NewSQLStore(dbConn, true)
+	if err != nil {
+		return nil, nil, err
+	} else if err := cm.ConsensusSetSubscribe(sqlStore, ccid, nil); err != nil {
+		return nil, nil, err
+	}
+
+	contractsDir := filepath.Join(dir, "contracts")
+	if err := os.MkdirAll(contractsDir, 0700); err != nil {
+		return nil, nil, err
+	}
+
+	objectsDir := filepath.Join(dir, "objects")
+	if err := os.MkdirAll(objectsDir, 0700); err != nil {
+		return nil, nil, err
+	}
+
+	cleanup := func() error {
+		errs := []error{
+			g.Close(),
+			cm.Close(),
+			tp.Close(),
+			sqlStore.Close(),
+		}
+		for _, err := range errs {
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+
+	b := bus.New(syncer{g, tp}, chainManager{cm}, txpool{tp}, w, sqlStore, sqlStore, sqlStore, sqlStore)
+	return b, cleanup, nil
+}
+
+func newWorker(cfg WorkerConfig, walletKey consensus.PrivateKey) (*worker.Worker, func() error, error) {
+	b := bus.NewClient(cfg.BusAddr, cfg.BusPassword)
+	workerKey := blake2b.Sum256(append([]byte("worker"), walletKey...))
+	w := worker.New(workerKey, b)
+	return w, func() error { return nil }, nil
+}
+
+func newAutopilot(cfg AutopilotConfig, dir string) (*autopilot.Autopilot, func() error, error) {
+	store, err := stores.NewJSONAutopilotStore(dir)
+	if err != nil {
+		return nil, nil, err
+	}
+	b := bus.NewClient(cfg.BusAddr, cfg.BusPassword)
+	w := worker.NewClient(cfg.WorkerAddr, cfg.WorkerPassword)
+	a, err := autopilot.New(store, b, w, cfg.LoopInterval)
+	if err != nil {
+		return nil, nil, err
+	}
+	cleanup := func() error {
+		a.Stop()
+		return nil
+	}
+	return a, cleanup, nil
+}
+
+// Node describes a single renterd process with one or more enabled components.
```

Comment:
Maybe every component should expose a management API that allows one node to register itself with another node. For instance, we could have the autopilot require a `bus` and expose a method called `registerWorker`, which would then in turn set the bus on the registering worker.

Reply:
I think that's a bit out-of-scope for this PR but I thought about something like that. But only on the bus since that should be our source of truth.
So the requirement would be that every node needs a bus but the bus can tell you about other workers and potentially autopilots that are connected to it.
Potentially through a web socket connection that is kept open between the bus and the other components. In the future we could use that to send events. Such as "hey a worker disconnected". Then the autopilot knows to not use that worker anymore.